diff --git a/subst.c b/subst.c
index 9f15f0b..7b57aec 100644
--- a/subst.c
+++ b/subst.c
@@ -7098,6 +7098,7 @@ parameter_brace_expand (string, indexp, quoted, pflags, quoted_dollar_atp, conta
       temp1 = parameter_brace_remove_pattern (name, temp, value, c, quoted);
       free (temp);
       free (value);
+      free (name);
 
       ret = alloc_word_desc ();
       ret->word = temp1;
-- 
2.5.5

