diff -up bash-4.1/sig.c.signal_sarestart bash-4.1/sig.c
--- bash-4.1/sig.c.signal_sarestart	2013-04-11 16:47:46.655592879 +0200
+++ bash-4.1/sig.c	2013-04-11 16:49:47.054185983 +0200
@@ -650,6 +650,15 @@ set_signal_handler (sig, handler)
 
   act.sa_handler = handler;
   act.sa_flags = 0;
+
+  /* vvv Copied from bash-4.2 vvv */
+  /* XXX - bash-4.2 */
+  /* We don't want a child death to interrupt interruptible system calls, even
+     if we take the time to reap children */
+  if (sig == SIGCHLD)
+    act.sa_flags |= SA_RESTART;         /* XXX */
+  /* ^^^ Copied from bash-4.2 ^^^ */
+
 #if 0
   if (sig == SIGALRM)
     act.sa_flags |= SA_INTERRUPT;	/* XXX */
