diff --git a/builtins/read.def b/builtins/read.def
--- a/builtins/read.def
+++ b/builtins/read.def
@@ -455,6 +455,7 @@ read_builtin (list)
      of the unwind-protect stack after the realloc() works right. */
   add_unwind_protect (xfree, input_string);
   interrupt_immediately++;
+  terminate_immediately++;
 
   unbuffered_read = (nchars > 0) || (delim != '\n') || input_is_pipe;
 
@@ -511,7 +512,6 @@ read_builtin (list)
 
       if (retval <= 0)
 	{
-	  CHECK_TERMSIG;
 	  eof = 1;
 	  break;
 	}
@@ -616,6 +616,7 @@ add_char:
     zsyncfd (fd);
 
   interrupt_immediately--;
+  terminate_immediately--;
   discard_unwind_frame ("read_builtin");
 
   retval = eof ? EXECUTION_FAILURE : EXECUTION_SUCCESS;
