--- bash-4.1-22/arrayfunc.c	2009-09-07 20:15:40.000000000 +0200
+++ bash-4.1/arrayfunc.c	2014-06-10 14:51:18.310014385 +0200
@@ -658,6 +658,7 @@ unbind_array_element (var, sub)
 	  return -1;
 	}
       assoc_remove (assoc_cell (var), akey);
+      free (akey);
     }
   else
     {
@@ -944,7 +945,10 @@ index_error:
       if (array_p (var) == 0 && assoc_p (var) == 0)
 	return (ind == 0 ? value_cell (var) : (char *)NULL);
       else if (assoc_p (var))
+      {
 	retval = assoc_reference (assoc_cell (var), akey);
+    free (akey);
+      }
       else
 	retval = array_reference (array_cell (var), ind);
     }
